File: processing/dji_photogrammetry/camera_trigger.py
"""
Camera trigger system for DJI drones with GPS/distance-based automation.

Handles automatic camera triggering based on time intervals, distance traveled,
or GPS coordinates. Logs all triggers with exact GPS position, altitude, 
drone attitude, timestamp, and camera settings.
"""

import time
import json
import math
from enum import Enum
from typing import Dict, Any, List, Optional, Tuple, Callable
from dataclasses import dataclass, asdict
from datetime import datetime


class TriggerMode(Enum):
    """Camera trigger modes."""
    TIME_BASED = "time"
    DISTANCE_BASED = "distance"
    WAYPOINT_BASED = "waypoint"
    MANUAL = "manual"


@dataclass
class DroneState:
    """Current drone state information."""
    latitude: float
    longitude: float
    altitude_m: float
    heading_deg: float
    pitch_deg: float
    roll_deg: float
    yaw_deg: float
    ground_speed_ms: float
    timestamp: datetime
    
    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary with ISO timestamp."""
        data = asdict(self)
        data['timestamp'] = self.timestamp.isoformat()
        return data


@dataclass
class CameraState:
    """Camera settings and EXIF information."""
    iso: int = 100
    aperture: str = "f/2.8"
    shutter_speed: str = "1/500"
    focal_length_mm: float = 24.0
    white_balance: str = "auto"
    exposure_mode: str = "auto"
    focus_mode: str = "infinity"
    image_format: str = "JPEG"
    
    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary."""
        return asdict(self)


@dataclass
class TriggerEvent:
    """Individual camera trigger event with complete metadata."""
    trigger_id: str
    trigger_mode: TriggerMode
    drone_state: DroneState
    camera_state: CameraState
    image_filename: Optional[str] = None
    trigger_reason: str = ""
    
    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary for JSON export."""
        return {
            'trigger_id': self.trigger_id,
            'trigger_mode': self.trigger_mode.value,
            'drone_state': self.drone_state.to_dict(),
            'camera_state': self.camera_state.to_dict(),
            'image_filename': self.image_filename,
            'trigger_reason': self.trigger_reason
        }


class CameraTrigger:
    """
    Camera trigger system for DJI drones.
    
    Manages automatic camera triggering based on configurable intervals
    and logs complete metadata for each capture.
    """
    
    def __init__(self):
        """Initialize camera trigger system."""
        self.trigger_events: List[TriggerEvent] = []
        self.last_trigger_time: Optional[datetime] = None
        self.last_trigger_position: Optional[Tuple[float, float]] = None
        self.trigger_count = 0
        self.is_active = False
        
        # Trigger settings
        self.time_interval_s: Optional[float] = None
        self.distance_interval_m: Optional[float] = None
        self.current_mode = TriggerMode.MANUAL
        
        # Callbacks for DJI SDK integration
        self.on_trigger_callback: Optional[Callable[[], str]] = None
        self.get_drone_state_callback: Optional[Callable[[], DroneState]] = None
        self.get_camera_state_callback: Optional[Callable[[], CameraState]] = None
    
    def set_trigger_mode(self, mode: TriggerMode, **kwargs):
        """
        Set trigger mode and parameters.
        
        Args:
            mode: Trigger mode (TIME_BASED, DISTANCE_BASED, etc.)
            **kwargs: Mode-specific parameters:
                - time_interval_s: for TIME_BASED mode
                - distance_interval_m: for DISTANCE_BASED mode
        """
        self.current_mode = mode
        
        if mode == TriggerMode.TIME_BASED:
            self.time_interval_s = kwargs.get('time_interval_s', 2.0)
            self.distance_interval_m = None
        elif mode == TriggerMode.DISTANCE_BASED:
            self.distance_interval_m = kwargs.get('distance_interval_m', 30.0)
            self.time_interval_s = None
        elif mode == TriggerMode.WAYPOINT_BASED:
            # Trigger only at specific waypoints
            self.time_interval_s = None
            self.distance_interval_m = None
        else:  # MANUAL
            self.time_interval_s = None
            self.distance_interval_m = None
    
    def set_callbacks(self, 
                     on_trigger: Callable[[], str],
                     get_drone_state: Callable[[], DroneState],
                     get_camera_state: Callable[[], CameraState]):
        """
        Set callback functions for DJI SDK integration.
        
        Args:
            on_trigger: Function to trigger camera, returns image filename
            get_drone_state: Function to get current drone state
            get_camera_state: Function to get current camera settings
        """
        self.on_trigger_callback = on_trigger
        self.get_drone_state_callback = get_drone_state
        self.get_camera_state_callback = get_camera_state
    
    def start_triggering(self):
        """Start automatic camera triggering."""
        self.is_active = True
        self.last_trigger_time = None
        self.last_trigger_position = None
        print(f"Camera triggering started in {self.current_mode.value} mode")
    
    def stop_triggering(self):
        """Stop automatic camera triggering."""
        self.is_active = False
        print(f"Camera triggering stopped. Total triggers: {self.trigger_count}")
    
    def update(self) -> bool:
        """
        Update trigger system with current drone state.
        
        Should be called regularly during flight to check trigger conditions.
        
        Returns:
            True if camera was triggered, False otherwise
        """
        if not self.is_active or not self.get_drone_state_callback:
            return False
        
        current_drone_state = self.get_drone_state_callback()
        current_time = datetime.now()
        
        should_trigger = False
        trigger_reason = ""
        
        if self.current_mode == TriggerMode.TIME_BASED:
            should_trigger, trigger_reason = self._check_time_trigger(current_time)
        elif self.current_mode == TriggerMode.DISTANCE_BASED:
            should_trigger, trigger_reason = self._check_distance_trigger(current_drone_state)
        elif self.current_mode == TriggerMode.WAYPOINT_BASED:
            # Waypoint triggering is handled externally
            should_trigger = False
        
        if should_trigger:
            return self.trigger_camera(current_drone_state, trigger_reason)
        
        return False
    
    def _check_time_trigger(self, current_time: datetime) -> Tuple[bool, str]:
        """Check if time-based trigger condition is met."""
        if self.time_interval_s is None:
            return False, ""
        
        if self.last_trigger_time is None:
            return True, f"Initial time-based trigger (interval: {self.time_interval_s}s)"
        
        time_elapsed = (current_time - self.last_trigger_time).total_seconds()
        if time_elapsed >= self.time_interval_s:
            return True, f"Time interval reached ({time_elapsed:.1f}s >= {self.time_interval_s}s)"
        
        return False, ""
    
    def _check_distance_trigger(self, drone_state: DroneState) -> Tuple[bool, str]:
        """Check if distance-based trigger condition is met."""
        if self.distance_interval_m is None:
            return False, ""
        
        current_pos = (drone_state.latitude, drone_state.longitude)
        
        if self.last_trigger_position is None:
            return True, f"Initial distance-based trigger (interval: {self.distance_interval_m}m)"
        
        distance_traveled = self._calculate_distance(
            self.last_trigger_position, current_pos
        )
        
        if distance_traveled >= self.distance_interval_m:
            return True, f"Distance interval reached ({distance_traveled:.1f}m >= {self.distance_interval_m}m)"
        
        return False, ""
    
    def _calculate_distance(self, pos1: Tuple[float, float], 
                          pos2: Tuple[float, float]) -> float:
        """
        Calculate distance between two GPS coordinates using Haversine formula.
        
        Returns:
            Distance in meters
        """
        lat1, lon1 = pos1
        lat2, lon2 = pos2
        
        # Convert to radians
        lat1_rad = math.radians(lat1)
        lon1_rad = math.radians(lon1)
        lat2_rad = math.radians(lat2)
        lon2_rad = math.radians(lon2)
        
        # Haversine formula
        dlat = lat2_rad - lat1_rad
        dlon = lon2_rad - lon1_rad
        
        a = (math.sin(dlat / 2) ** 2 + 
             math.cos(lat1_rad) * math.cos(lat2_rad) * math.sin(dlon / 2) ** 2)
        c = 2 * math.asin(math.sqrt(a))
        
        # Earth radius in meters
        earth_radius_m = 6371000
        return earth_radius_m * c
    
    def trigger_camera(self, drone_state: DroneState = None, 
                      reason: str = "Manual trigger") -> bool:
        """
        Trigger camera capture and log metadata.
        
        Args:
            drone_state: Current drone state (if None, will get from callback)
            reason: Reason for trigger (for logging)
            
        Returns:
            True if trigger was successful, False otherwise
        """
        if not self.on_trigger_callback:
            print("Warning: No trigger callback set")
            return False
        
        # Get current states
        if drone_state is None and self.get_drone_state_callback:
            drone_state = self.get_drone_state_callback()
        
        if self.get_camera_state_callback:
            camera_state = self.get_camera_state_callback()
        else:
            camera_state = CameraState()  # Default settings
        
        if drone_state is None:
            print("Warning: No drone state available")
            return False
        
        # Trigger camera
        try:
            image_filename = self.on_trigger_callback()
        except Exception as e:
            print(f"Camera trigger failed: {e}")
            return False
        
        # Create trigger event
        self.trigger_count += 1
        trigger_id = f"trigger_{self.trigger_count:06d}_{int(time.time())}"
        
        trigger_event = TriggerEvent(
            trigger_id=trigger_id,
            trigger_mode=self.current_mode,
            drone_state=drone_state,
            camera_state=camera_state,
            image_filename=image_filename,
            trigger_reason=reason
        )
        
        self.trigger_events.append(trigger_event)
        
        # Update trigger history
        self.last_trigger_time = drone_state.timestamp
        self.last_trigger_position = (drone_state.latitude, drone_state.longitude)
        
        print(f"Camera triggered: {trigger_id} - {reason}")
        return True
    
    def trigger_at_waypoint(self, waypoint_id: str, drone_state: DroneState = None) -> bool:
        """
        Trigger camera at specific waypoint.
        
        Args:
            waypoint_id: Identifier for the waypoint
            drone_state: Current drone state
            
        Returns:
            True if trigger was successful
        """
        reason = f"Waypoint trigger: {waypoint_id}"
        return self.trigger_camera(drone_state, reason)
    
    def export_metadata_csv(self, filename: str):
        """
        Export trigger metadata to CSV file.
        
        Args:
            filename: Output CSV filename
        """
        import pandas as pd
        
        if not self.trigger_events:
            print("No trigger events to export")
            return
        
        # Flatten trigger events for CSV export
        csv_data = []
        for event in self.trigger_events:
            row = {
                'trigger_id': event.trigger_id,
                'trigger_mode': event.trigger_mode.value,
                'image_filename': event.image_filename,
                'trigger_reason': event.trigger_reason,
                
                # Drone state
                'latitude': event.drone_state.latitude,
                'longitude': event.drone_state.longitude,
                'altitude_m': event.drone_state.altitude_m,
                'heading_deg': event.drone_state.heading_deg,
                'pitch_deg': event.drone_state.pitch_deg,
                'roll_deg': event.drone_state.roll_deg,
                'yaw_deg': event.drone_state.yaw_deg,
                'ground_speed_ms': event.drone_state.ground_speed_ms,
                'timestamp': event.drone_state.timestamp.isoformat(),
                
                # Camera state
                'iso': event.camera_state.iso,
                'aperture': event.camera_state.aperture,
                'shutter_speed': event.camera_state.shutter_speed,
                'focal_length_mm': event.camera_state.focal_length_mm,
                'white_balance': event.camera_state.white_balance,
                'exposure_mode': event.camera_state.exposure_mode,
                'focus_mode': event.camera_state.focus_mode,
                'image_format': event.camera_state.image_format,
            }
            csv_data.append(row)
        
        df = pd.DataFrame(csv_data)
        df.to_csv(filename, index=False)
        print(f"Exported {len(csv_data)} trigger events to {filename}")
    
    def export_metadata_json(self, filename: str):
        """
        Export trigger metadata to JSON file.
        
        Args:
            filename: Output JSON filename
        """
        if not self.trigger_events:
            print("No trigger events to export")
            return
        
        export_data = {
            'mission_info': {
                'total_triggers': len(self.trigger_events),
                'trigger_mode': self.current_mode.value,
                'export_timestamp': datetime.now().isoformat()
            },
            'trigger_events': [event.to_dict() for event in self.trigger_events]
        }
        
        with open(filename, 'w') as f:
            json.dump(export_data, f, indent=2)
        
        print(f"Exported {len(self.trigger_events)} trigger events to {filename}")
    
    def get_mission_summary(self) -> Dict[str, Any]:
        """
        Get summary statistics for current mission.
        
        Returns:
            Dictionary with mission statistics
        """
        if not self.trigger_events:
            return {'total_triggers': 0}
        
        # Calculate mission duration
        first_trigger = self.trigger_events[0].drone_state.timestamp
        last_trigger = self.trigger_events[-1].drone_state.timestamp
        duration_s = (last_trigger - first_trigger).total_seconds()
        
        # Calculate total distance traveled
        total_distance = 0
        for i in range(1, len(self.trigger_events)):
            pos1 = (self.trigger_events[i-1].drone_state.latitude,
                   self.trigger_events[i-1].drone_state.longitude)
            pos2 = (self.trigger_events[i].drone_state.latitude,
                   self.trigger_events[i].drone_state.longitude)
            total_distance += self._calculate_distance(pos1, pos2)
        
        # Get altitude range
        altitudes = [event.drone_state.altitude_m for event in self.trigger_events]
        
        return {
            'total_triggers': len(self.trigger_events),
            'mission_duration_s': duration_s,
            'total_distance_m': total_distance,
            'trigger_mode': self.current_mode.value,
            'altitude_range_m': {
                'min': min(altitudes) if altitudes else 0,
                'max': max(altitudes) if altitudes else 0,
                'avg': sum(altitudes) / len(altitudes) if altitudes else 0
            },
            'first_trigger': first_trigger.isoformat(),
            'last_trigger': last_trigger.isoformat()
        }

File: mobile/android-sdk-wrapper/src/main/java/com/dji/photogrammetry/CameraTrigger.kt
package com.dji.photogrammetry

import com.google.gson.Gson
import com.google.gson.annotations.SerializedName
import java.util.*
import kotlin.math.*

/**
 * Camera trigger system for DJI drones with GPS/distance-based automation.
 * 
 * Handles automatic camera triggering based on time intervals, distance traveled,
 * or GPS coordinates. Logs all triggers with exact GPS position, altitude,
 * drone attitude, timestamp, and camera settings.
 */

enum class TriggerMode {
    @SerializedName("time")
    TIME_BASED,
    @SerializedName("distance") 
    DISTANCE_BASED,
    @SerializedName("waypoint")
    WAYPOINT_BASED,
    @SerializedName("manual")
    MANUAL
}

data class DroneState(
    val latitude: Double,
    val longitude: Double,
    val altitudeM: Double,
    val headingDeg: Double,
    val pitchDeg: Double,
    val rollDeg: Double,
    val yawDeg: Double,
    val groundSpeedMs: Double,
    val timestamp: Date
)

data class CameraState(
    val iso: Int = 100,
    val aperture: String = "f/2.8",
    val shutterSpeed: String = "1/500",
    val focalLengthMm: Double = 24.0,
    val whiteBalance: String = "auto",
    val exposureMode: String = "auto",
    val focusMode: String = "infinity",
    val imageFormat: String = "JPEG"
)

data class TriggerEvent(
    val triggerId: String,
    val triggerMode: TriggerMode,
    val droneState: DroneState,
    val cameraState: CameraState,
    val imageFilename: String? = null,
    val triggerReason: String = ""
)

data class MissionSummary(
    val totalTriggers: Int,
    val missionDurationS: Long,
    val totalDistanceM: Double,
    val triggerMode: String,
    val altitudeRange: AltitudeRange,
    val firstTrigger: Date?,
    val lastTrigger: Date?
)

data class AltitudeRange(
    val min: Double,
    val max: Double,
    val avg: Double
)

interface CameraTriggerListener {
    fun onCameraTrigger(): String? // Returns image filename or null if failed
    fun getCurrentDroneState(): DroneState?
    fun getCurrentCameraState(): CameraState?
}

class CameraTrigger {
    private val triggerEvents = mutableListOf<TriggerEvent>()
    private var lastTriggerTime: Date? = null
    private var lastTriggerPosition: Pair<Double, Double>? = null
    private var triggerCount = 0
    private var isActive = false
    
    // Trigger settings
    private var timeIntervalS: Double? = null
    private var distanceIntervalM: Double? = null
    private var currentMode = TriggerMode.MANUAL
    
    // Listener for DJI SDK integration
    private var listener: CameraTriggerListener? = null
    
    fun setTriggerMode(mode: TriggerMode, vararg params: Pair<String, Any>) {
        currentMode = mode
        
        when (mode) {
            TriggerMode.TIME_BASED -> {
                timeIntervalS = params.find { it.first == "timeIntervalS" }?.second as? Double ?: 2.0
                distanceIntervalM = null
            }
            TriggerMode.DISTANCE_BASED -> {
                distanceIntervalM = params.find { it.first == "distanceIntervalM" }?.second as? Double ?: 30.0
                timeIntervalS = null
            }
            TriggerMode.WAYPOINT_BASED, TriggerMode.MANUAL -> {
                timeIntervalS = null
                distanceIntervalM = null
            }
        }
    }
    
    fun setListener(listener: CameraTriggerListener) {
        this.listener = listener
    }
    
    fun startTriggering() {
        isActive = true
        lastTriggerTime = null
        lastTriggerPosition = null
        println("Camera triggering started in ${currentMode.name} mode")
    }
    
    fun stopTriggering() {
        isActive = false
        println("Camera triggering stopped. Total triggers: $triggerCount")
    }
    
    /**
     * Update trigger system with current drone state.
     * Should be called regularly during flight to check trigger conditions.
     */
    fun update(): Boolean {
        if (!isActive || listener == null) return false
        
        val currentDroneState = listener?.getCurrentDroneState() ?: return false
        val currentTime = Date()
        
        val (shouldTrigger, triggerReason) = when (currentMode) {
            TriggerMode.TIME_BASED -> checkTimeTrigger(currentTime)
            TriggerMode.DISTANCE_BASED -> checkDistanceTrigger(currentDroneState)
            else -> Pair(false, "")
        }
        
        return if (shouldTrigger) {
            triggerCamera(currentDroneState, triggerReason)
        } else false
    }
    
    private fun checkTimeTrigger(currentTime: Date): Pair<Boolean, String> {
        val interval = timeIntervalS ?: return Pair(false, "")
        
        val lastTime = lastTriggerTime
        if (lastTime == null) {
            return Pair(true, "Initial time-based trigger (interval: ${interval}s)")
        }
        
        val timeElapsed = (currentTime.time - lastTime.time) / 1000.0
        return if (timeElapsed >= interval) {
            Pair(true, "Time interval reached (${timeElapsed}s >= ${interval}s)")
        } else {
            Pair(false, "")
        }
    }
    
    private fun checkDistanceTrigger(droneState: DroneState): Pair<Boolean, String> {
        val interval = distanceIntervalM ?: return Pair(false, "")
        
        val currentPos = Pair(droneState.latitude, droneState.longitude)
        val lastPos = lastTriggerPosition
        
        if (lastPos == null) {
            return Pair(true, "Initial distance-based trigger (interval: ${interval}m)")
        }
        
        val distanceTraveled = calculateDistance(lastPos, currentPos)
        return if (distanceTraveled >= interval) {
            Pair(true, "Distance interval reached (${distanceTraveled}m >= ${interval}m)")
        } else {
            Pair(false, "")
        }
    }
    
    private fun calculateDistance(pos1: Pair<Double, Double>, pos2: Pair<Double, Double>): Double {
        val (lat1, lon1) = pos1
        val (lat2, lon2) = pos2
        
        // Convert to radians
        val lat1Rad = Math.toRadians(lat1)
        val lon1Rad = Math.toRadians(lon1)
        val lat2Rad = Math.toRadians(lat2)
        val lon2Rad = Math.toRadians(lon2)
        
        // Haversine formula
        val dlat = lat2Rad - lat1Rad
        val dlon = lon2Rad - lon1Rad
        
        val a = sin(dlat / 2).pow(2) + cos(lat1Rad) * cos(lat2Rad) * sin(dlon / 2).pow(2)
        val c = 2 * asin(sqrt(a))
        
        // Earth radius in meters
        val earthRadiusM = 6371000.0
        return earthRadiusM * c
    }
    
    /**
     * Trigger camera capture and log metadata.
     */
    fun triggerCamera(droneState: DroneState? = null, reason: String = "Manual trigger"): Boolean {
        val currentListener = listener ?: return false
        
        // Get current states
        val currentDroneState = droneState ?: currentListener.getCurrentDroneState() ?: return false
        val cameraState = currentListener.getCurrentCameraState() ?: CameraState()
        
        // Trigger camera
        val imageFilename = try {
            currentListener.onCameraTrigger()
        } catch (e: Exception) {
            println("Camera trigger failed: ${e.message}")
            return false
        }
        
        // Create trigger event
        triggerCount++
        val triggerId = "trigger_${triggerCount.toString().padStart(6, '0')}_${System.currentTimeMillis()}"
        
        val triggerEvent = TriggerEvent(
            triggerId = triggerId,
            triggerMode = currentMode,
            droneState = currentDroneState,
            cameraState = cameraState,
            imageFilename = imageFilename,
            triggerReason = reason
        )
        
        triggerEvents.add(triggerEvent)
        
        // Update trigger history
        lastTriggerTime = currentDroneState.timestamp
        lastTriggerPosition = Pair(currentDroneState.latitude, currentDroneState.longitude)
        
        println("Camera triggered: $triggerId - $reason")
        return true
    }
    
    /**
     * Trigger camera at specific waypoint.
     */
    fun triggerAtWaypoint(waypointId: String, droneState: DroneState? = null): Boolean {
        val reason = "Waypoint trigger: $waypointId"
        return triggerCamera(droneState, reason)
    }
    
    /**
     * Export trigger metadata to JSON string.
     */
    fun exportMetadataJson(): String {
        val exportData = mapOf(
            "mission_info" to mapOf(
                "total_triggers" to triggerEvents.size,
                "trigger_mode" to currentMode.name,
                "export_timestamp" to Date().toString()
            ),
            "trigger_events" to triggerEvents
        )
        
        return Gson().toJson(exportData)
    }
    
    /**
     * Get mission summary statistics.
     */
    fun getMissionSummary(): MissionSummary {
        if (triggerEvents.isEmpty()) {
            return MissionSummary(
                totalTriggers = 0,
                missionDurationS = 0,
                totalDistanceM = 0.0,
                triggerMode = currentMode.name,
                altitudeRange = AltitudeRange(0.0, 0.0, 0.0),
                firstTrigger = null,
                lastTrigger = null
            )
        }
        
        // Calculate mission duration
        val firstTrigger = triggerEvents.first().droneState.timestamp
        val lastTrigger = triggerEvents.last().droneState.timestamp
        val durationS = (lastTrigger.time - firstTrigger.time) / 1000
        
        // Calculate total distance traveled
        var totalDistance = 0.0
        for (i in 1 until triggerEvents.size) {
            val pos1 = Pair(
                triggerEvents[i - 1].droneState.latitude,
                triggerEvents[i - 1].droneState.longitude
            )
            val pos2 = Pair(
                triggerEvents[i].droneState.latitude,
                triggerEvents[i].droneState.longitude
            )
            totalDistance += calculateDistance(pos1, pos2)
        }
        
        // Get altitude range
        val altitudes = triggerEvents.map { it.droneState.altitudeM }
        val altitudeRange = AltitudeRange(
            min = altitudes.minOrNull() ?: 0.0,
            max = altitudes.maxOrNull() ?: 0.0,
            avg = altitudes.average()
        )
        
        return MissionSummary(
            totalTriggers = triggerEvents.size,
            missionDurationS = durationS,
            totalDistanceM = totalDistance,
            triggerMode = currentMode.name,
            altitudeRange = altitudeRange,
            firstTrigger = firstTrigger,
            lastTrigger = lastTrigger
        )
    }
    
    /**
     * Clear all trigger events and reset counters.
     */
    fun reset() {
        triggerEvents.clear()
        triggerCount = 0
        lastTriggerTime = null
        lastTriggerPosition = null
        isActive = false
    }
    
    /**
     * Get all trigger events.
     */
    fun getTriggerEvents(): List<TriggerEvent> = triggerEvents.toList()
}